hpx/parallel/algorithms/merge.hpp

```diff
@@ -491,6 +490,341 @@ namespace hpx { namespace parallel { inline namespace v1
+                            comp, proj);
+                    });
+
+                try {
+                    // Process the range which is rightside of 'target'.
+                    parallel_inplace_merge_helper(policy,
+                        target + 1, pivot + 1, last,
+                        comp, proj);
+                }
+                catch (...) {
+                    fut.wait();
+
+                    std::vector<hpx::future<void>> futures(2);
+                    futures[0] = std::move(fut);
+                    futures[1] = hpx::make_exceptional_future<void>(
+                        std::current_exception());
+
+                    std::list<std::exception_ptr> errors;
+                    util::detail::handle_local_exceptions<ExPolicy>::call(
+                        futures, errors);
+
+                    // Not reachable.
+                    HPX_ASSERT(false);
+                }
+
+                fut.get();
+            }
+        }
+
+        template <typename ExPolicy, typename RandIter,
+            typename Comp, typename Proj,
+        HPX_CONCEPT_REQUIRES_(
+            hpx::traits::is_random_access_iterator<RandIter>::value)>
+        inline RandIter
+        parallel_inplace_merge(ExPolicy policy,
+            RandIter first, RandIter middle, RandIter last,
+            Comp && comp, Proj && proj)
+        {
+            parallel_inplace_merge_helper(
+                std::forward<ExPolicy>(policy),
+                first, middle, last,
+                std::forward<Comp>(comp),
+                std::forward<Proj>(proj));
+
+            return last;
+        }
+
+        template <typename Iter>
+        struct inplace_merge : public detail::algorithm<inplace_merge<Iter>, Iter>
+        {
+            inplace_merge()
+              : inplace_merge::algorithm("inplace_merge")
+            {}
+
+            template <typename ExPolicy, typename RandIter,
+                typename Comp, typename Proj>
+            static RandIter
+            sequential(ExPolicy,
+                RandIter first, RandIter middle, RandIter last,
+                Comp && comp, Proj && proj)
+            {
+                return sequential_inplace_merge(
+                    first, middle, last,
+                    std::forward<Comp>(comp),
+                    std::forward<Proj>(proj));
+            }
+
+            template <typename ExPolicy, typename RandIter,
+                typename Comp, typename Proj>
+            static typename util::detail::algorithm_result<
+                ExPolicy, RandIter
+            >::type
+            parallel(ExPolicy && policy,
+                RandIter first, RandIter middle, RandIter last,
+                Comp && comp, Proj && proj)
+            {
+                typedef util::detail::algorithm_result<
+                    ExPolicy, RandIter
+                > algorithm_result;
+
+                try {
+                    return algorithm_result::get(
```

Comment:
If implementing a asynchronous version of `inplace_merge` is not feasible/necessary, could you at least run the algorithm on a new thread for `par(task)` and similar?

Reply:
I was just wondering. No need to change things.

Reply:
@hkaiser Oh.. I was very fool.... :(
I don't know why I didn't consider asynchronous execution policy..
As you say, I modify the algorithm will run on a new thread.
There are similar problems in parallel::merge and parallel::partition, too.
I will fix them in another PR soon.

Reply:
@hkaiser The problem you suggested should be fixed as I think. Why I don't need to change things?

Reply:
What I meant is that there is no need to remove the concept checks. 
I'd very much appreciate you adding the asynchronous behavior, though.